### ISSUE TITLE
Admin: Fix the usage of userType

### ISSUE THREAD

**Issue by palisadoes** [PalisadoesFoundation/talawa-admin#1445]
### NOTE

1. This issue will require close coordination with persons assigned issues in the reference section.
2. You may find it easier to split responsibilities in a different way than proposed here to be more efficient is creating the goal

**Describe the bug**

1. The schema doesn't apply a person's `userType` by organization. 
1. The `userType` is currently a part of the `User` definition not part of a `userOrganization` definition or something similar. 
    1. This means that an `Admin` is an `Admin` for all organizations which grants the same rights as a `SuperAdmin`. This is not the intention.
    2. This has created historical confusion in the `Admin` and `SuperAdmin` roles.

This needs to be resolved as it is a showstopper bug.

**Expected behavior**

These conditions must apply:

1. The `userType` must be applied per person per organization.
2. Each `SuperAdmin` must be a `SuperAdmin` for all organizations
3. All other `userType`s would be on a per organization basis

This means that all code logic referencing the `userType` will need to be updated accordingly.

_TALAWA API_

Additional considerations include:

1. New user registrations will need to default to the `userType` `USER` if not done so already
1. The Schema will need to be updated to reflect this requirement
1. Mutations and queries will need to be updated and in some cases new ones added.

_TALAWA ADMIN_

Additional considerations include:

1. The current screen that edits a person's userType will need to be updated. A design is being created for this and will be reflected in the Figma file on our Design Guide page:
    1. https://docs.talawa.io/docs/design/ux/ux-design-system

_TALAWA MOBILE_

No expected changes as userType isn't referenced in the code base

**Actual behavior**

- The schema doesn't apply a person's `userType` by organization. 

**Screenshots**

The current schema:

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/cbd791b2-3ff7-44b9-b936-023c9d979b01)

**Additional details**

Related issues:

1. https://github.com/PalisadoesFoundation/talawa-api/issues/1711
1. https://github.com/PalisadoesFoundation/talawa/issues/2373
1. https://github.com/PalisadoesFoundation/talawa-admin/issues/1445
1. https://github.com/PalisadoesFoundation/talawa-admin/issues/1451
1. https://github.com/PalisadoesFoundation/talawa-admin/issues/1452
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1862
1. https://github.com/PalisadoesFoundation/talawa-api/issues/1904

**Potential internship candidates**

Please read this if you are planning to apply for a Palisadoes Foundation internship 
- https://github.com/PalisadoesFoundation/talawa/issues/359

**Comment by KoraboinaSankeerth** [comment]
Would like to work on this issue.

**Comment by Sejal1411** [comment]
would like to work on this issue! Pls assign

**Comment by manishjha-04** [comment]
would like to work in this issue!

**Comment by NayOoLwin5** [comment]
Would like to work on this issue.

**Comment by Pranavx1** [comment]
Can i work on this?

**Comment by palisadoes** [comment]
@KoraboinaSankeerth 

1. We have created profile design issues that will impact your work. We expect a delay in the API work while we finalize the methodology. This will mean the design issues should be either completed or near completion by the time you are ready.
2. The issues are referenced in the first comment

**Comment by palisadoes** [comment]
Unassigning due to no PR activity or assignee comments in the past 10 days

**Comment by NayOoLwin5** [comment]
I would like to work on this issue

**Comment by Cioppolo14** [comment]
@NayOoLwin5 You currently have 2 issues. Our policy is to assign no more than two issues to each contributor across all repositories. This way everyone gets a chance to participate in the projects. We sometimes give exceptions for more urgent cases and sometimes we lose track, but the policy stands. You have reached your limit, please wait until your existing issues are closed before requesting more issues. You could unassign yourself from one of the other issues too.

**Comment by KoraboinaSankeerth** [comment]
Hey,I currently have semester end examinations,assign me the task.I would
be able to work on the issue from 2 feb

**Comment by manishjha-04** [comment]
Hey can I work on this issue?

**Comment by manishjha-04** [comment]
Hey can I work on this issue?

**Comment by git-init-priyanshu** [comment]
I would like to work on this issue. Thanks.

**Comment by Cioppolo14** [comment]
@KoraboinaSankeerth I'm going to give someone else a chance with this issue.  When your exams are done, please check back for open issues, we would love to collaborate with you.

**Comment by git-init-priyanshu** [comment]
I need to check the userType of the User which could be either of these USER, NON_USER, ADMIN, SUPER_ADMIN and I need to change the code accordingly. Am I right?

**Comment by palisadoes** [comment]
The issue description covers what you need to do. Read it carefully

**Comment by git-init-priyanshu** [comment]
Unassigning myself.
As much as I love to work on this issue, I'm currently unable to allocate time for it. I would like to work on this in the future when I have a bit more availability. Meanwhile, I encourage someone else can try to fix this issue. Thanks.

**Comment by adi790uu** [comment]
Would like to solve this @Cioppolo14

**Comment by Olatade** [comment]
@adi790uu are you still working on this ?

**Comment by adi790uu** [comment]
@Olatade I am waiting for issue in the api section to be merged.

**Comment by palisadoes** [comment]
@adi790uu @Olatade @AVtheking

1. The API PR authored by @AVtheking was just merged. 
    1. https://github.com/PalisadoesFoundation/talawa-api/issues/1711
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1830
1. We created a new API branch for this feature called `develop-userTypeFix`
1. We have created a Talawa-Admin feature branch for this called `develop-userTypeFix` too. The PR for this issue must be made against this new branch.
1. We will find bugs. There are over 80 references to `userType` in the code base which will require coordination between API and Admin to fix. There will be PR merges to `develop` that will also need to be included.
    ![image](https://github.com/PalisadoesFoundation/talawa-admin/assets/16875803/3be627c4-c4bc-46ad-9a47-797d8dd5668a)
1. It will be very important to submit incremental PRs with small batches of updates. That will make PR review faster. So I'd suggest trying to do the changes one screen at a time, starting from login / registration. @adi790uu @AVtheking is this possible in your opinions?

This is a very important feature that will finally fix a long outstanding failure in the logic of our code.

**Comment by AVtheking** [comment]
@palisadoes sir Instead of one screen at a time I would suggest  we can do changes feature wise like auth,creating organisation,Action items ,etc

**Comment by palisadoes** [comment]
1. That would mean we'd need to create multiple smaller issues with this one as the parent.
2. Please work with @adi790uu, maybe on slack, to finalize what those smaller child issues would be and then create them. 
3. Think of ways in which they could be worked on in parallel to make implementation faster by delegating the work.
4. When the list is done, post them here so that we can discuss them
5. The PRs need to be very small so that the reviewers can quickly do their work. Many of them will only be available to review on weekends. Having a lot of small easy to review PRs will be the best approach.

Is this approach feasible?

**Comment by AVtheking** [comment]
Sir this would also need to be addressed in the mobile repo because there are more fields that are shifted from userModel to appUserProfile model such as adminFor,CreatedOrganization,eventAdmin,etc .These need to be changed accordingly

**Comment by palisadoes** [comment]
@adi790uu 

1. What approach are you planning?
1. Have you managed to coordinate with the others?

**Comment by adi790uu** [comment]
> @adi790uu 
> 
> 1. What approach are you planning?
> 1. Have you managed to coordinate with the others?

Yeah I have been talking with @AVtheking and identifying the potential changes.Will post them here soon.

**Comment by adi790uu** [comment]
@palisadoes 
1. We need to start with auth as LOGIN and SIGNUP mutation will be changed:

<img width="638" alt="Screenshot 2024-02-18 at 8 10 18 PM" src="https://github.com/PalisadoesFoundation/talawa-admin/assets/132701661/d987588c-b3a5-4c35-a2f6-f2cb82b75095">

2. We have to update the Users page as it updated the roles of Users, so changes in queries and mutations is required:

<img width="1470" alt="Screenshot 2024-02-18 at 8 03 18 PM" src="https://github.com/PalisadoesFoundation/talawa-admin/assets/132701661/31eebd56-2954-4ac7-a008-84d25fa4b27f">


do give your feedback on this I will list more changes which will be required soon.

**Comment by palisadoes** [comment]
1. Thanks for the feedback
2. I'd prefer if you take the initiative to do the updates based on the feedback you have had with   @AVtheking @Dante291.
3. Work on ways to submit a series of small PRs that are testable starting at a login / registration screen.
4. It's acceptable if some screens break with the PRs, as long as the series of screens updated in the PR function.

**Comment by palisadoes** [comment]
@adi790uu 

Be aware of this issue. Its PR will also need to be merged into the new branch too.

- https://github.com/PalisadoesFoundation/talawa-api/issues/1862

**Comment by adi790uu** [comment]
> 1. Admins are per organization based ,so user would be admin for the organization which are inside the `adminFor `filed ,if there are no organization inside the `adminFor `field then it means that the user is not admin for any organization.
> 2. `adminApproved `is used to access the admin web portal by a user

Okay @AVtheking got it now.

**Comment by palisadoes** [comment]
@adi790uu 

1. You will need to merge the latest `develop` branch into `develop-userTypeFix`
2. Apply this change to the `develop-userTypeFix` branch. 
3. The API already has a `develop-userTypeFix` branch that you can test against.

**Comment by palisadoes** [comment]
A `adminApproved` flag that is set authorizes a user to login the Talawa-Admin user screens

**Comment by palisadoes** [comment]
@adi790uu 
@AVtheking 
@Dante291

Are we ready to start merging `develop-userTypeFix` into `develop` now?

**Comment by Dante291** [comment]
> @adi790uu @AVtheking @Dante291
> 
> Are we ready to start merging `develop-userTypeFix` into `develop` now?

I have a raised a PR making major logic changes just have to update some queries and mutation in mobile repo.

**Comment by AVtheking** [comment]
> @adi790uu @AVtheking @Dante291
> 
> Are we ready to start merging `develop-userTypeFix` into `develop` now?

No sir rn only auth is changed,many queries and mutation are left

**Comment by adi790uu** [comment]
@palisadoes lot of work is still left

**Comment by Dante291** [comment]
> > @adi790uu @AVtheking @Dante291
> > Are we ready to start merging `develop-userTypeFix` into `develop` now?
> 
> No sir rn only auth is changed,many queries and mutation are left

@AVtheking in mobile only queries and mutations need to be updated now if I am not wrong...

**Comment by adi790uu** [comment]
logically changes need to be made in most of the screens so that they can work.

**Comment by adi790uu** [comment]
> > > @adi790uu @AVtheking @Dante291
> > > Are we ready to start merging `develop-userTypeFix` into `develop` now?
> > 
> > 
> > No sir rn only auth is changed,many queries and mutation are left
> 
> @AVtheking in mobile only queries and mutations need to be updated now if I am not wrong...

I didnt get your point.

**Comment by Dante291** [comment]
> > > > @adi790uu @AVtheking @Dante291
> > > > Are we ready to start merging `develop-userTypeFix` into `develop` now?
> > > 
> > > 
> > > No sir rn only auth is changed,many queries and mutation are left
> > 
> > 
> > @AVtheking in mobile only queries and mutations need to be updated now if I am not wrong...
> 
> I didnt get your point.

In Talawa mobile App I have made logical changes according to how now we are saving User details in the app now just some queries need to updated and then we will be good merge it with develop mobile branch

**Comment by adi790uu** [comment]
okay but work is still left in this repo.

**Comment by adi790uu** [comment]
@palisadoes we need to reopen this.

**Comment by palisadoes** [comment]
> > > > > @adi790uu @AVtheking @Dante291
> > > > > Are we ready to start merging `develop-userTypeFix` into `develop` now?
> > > > 
> > > > 
> > > > No sir rn only auth is changed,many queries and mutation are left
> > > 
> > > 
> > > @AVtheking in mobile only queries and mutations need to be updated now if I am not wrong...
> > 
> > 
> > I didnt get your point.
> 
> In Talawa mobile App I have made logical changes according to how we are saving User details in the app now just some queries need to updated and then we will be good merge it with develop mobile branch

Your most recent PR was merged. Please open the next one in the set

**Comment by palisadoes** [comment]
Adding @AdityaRaimec22 to this thread for context

**Comment by palisadoes** [comment]
> @palisadoes Yeah I have changed the logic in orgList screen but some tests are failing I have to fix those and then I will create a PR for that.

Is there a lot of work outstanding to begin the merge back into `develop`?

**Comment by adi790uu** [comment]
@palisadoes yes for each screen I have to fix the tests as well as some logical changes are made in the codebase, so it will take some time.

**Comment by AVtheking** [comment]
@palisadoes admin repo extensively used user type field and other which are moved to appUserProfile ,it will require lot of changes

**Comment by AVtheking** [comment]
@Dante291 after changing existing queries and mutation you have to lookout for the new merges in the mobile repo which woul require change accordingly

**Comment by AVtheking** [comment]
@adi790uu I would suggest you list down the screens which are decoupled from one another so that we can open issues and multiple people can work parallely on this

**Comment by palisadoes** [comment]
That would be the best approach.

@adi790uu Please create a list for this and post it here. We can use it too create a more efficient approach

**Comment by adi790uu** [comment]
@palisadoes @AVtheking Okay I will create a list of changes required and post it here today.

**Comment by palisadoes** [comment]
> @palisadoes @AVtheking Okay I will create a list of changes required and post it here today.

Please remember this so we can share the load

**Comment by adi790uu** [comment]
The changes will be done in mutations, queries, mocks, tests and components used in these screens mainly.

list of screens which require changes : 

1.MemberDetail
2.OrganizationPeople
3.OrgList
4.UserPortal/Home.tsx
5.UserPortal/Organizations
6.UserPortal/Users

there might be some screens where changes might not be required as some queries and mutations are used in one or more pages. @palisadoes

**Comment by adi790uu** [comment]
@palisadoes I have been working on OrgList the tests for this screen are taking time to be resolved but as soon as it is done will create a PR for it.

**Comment by adi790uu** [comment]
@palisadoes I have tried fixing the tests for Org List and it just doesn't work, will it be fine if I leave the tests and create a PR and move on to other components and at  last comeback to the tests which are consuming a lot of time fixing?

**Comment by palisadoes** [comment]
1. Create a separate issue for it and ask one of the others to work on its tests.
1. We need the tests for all the screens to be done. 
1. Create issues for some of the other screens so others can assist as suggested before. They can work on the tests for those screens independently

Historically the code gets bad very quickly when we ignore testing. This is an are in which we cannot compromise
